applications/asset_tracker_v2/src/modules/cloud_module.c

```diff
@@ -337,6 +349,36 @@ static void cloud_wrap_event_handler(const struct cloud_wrap_event *const evt)
 
 		int err;
 
+		if (IS_ENABLED(CONFIG_DATA_SAMPLE_LOCATION_DEFAULT)) {
+			/* TODO: This is now nRF cloud specific and must be made generic
+			 * if we go this way of passing the location to Location library,
+			 * and again back from there to ATv2.
+			 * We likely need a new event type for cellular location
+			 * (CLOUD_WRAP_EVT_CELLULAR_LOCATION_RECEIVED etc) and parsing
+			 * done in cloud integration files.
+			 */
+			err = cell_pos_response_filter(evt->data.buf, evt->data.len);
+			if (err) {
+				struct nrf_cloud_cell_pos_result result = {0};
+				struct location_data location_result;
+
+				err = nrf_cloud_cell_pos_process(evt->data.buf, &result);
+
+				// TODO: Check err = 1
+				if (err == 0) {
+					location_result.accuracy = (double)result.unc;
+					location_result.latitude = result.lat;
+					location_result.longitude = result.lon;
+					location_result.datetime.valid = false;
+					location_cellular_ready(&location_result);
```

Comment:
I'm thinking that If ncell measurements has been carried out and sent the "location" has been obtained. Why fallback from there?
I can imagine that there are a lot of application that only cares about sending the ncell measurements from the device and doesn't want to send back the location for energy reasons.

And. Lets say a customer wants to implement its own custom cloud solution and decides to resolve the ncell measurements themselves. Then, in order to properly implement the behavior of the location API it must sent the location back.

Could there be an option that when enabled calls `location_cellular_ready(NULL)` immediately after the location has been handed off to the system. From the location APIs standpoint it has done its job and its up to the application to handle the ncell measurements from there.

Reply:
> I'm thinking that If ncell measurements has been carried out and sent the "location" has been obtained. Why fallback from there?

There is no fallback if location is obtained based on ncell measurements. However, we don't know if location can be resolved from the ncell measurements until the cloud responds a successful response.

> Could there be an option that when enabled calls `location_cellular_ready(NULL)` immediately after the location has been handed off to the system. From the location APIs standpoint it has done its job and its up to the application to handle the ncell measurements from there.

This is an option I've had in my mind and we can digest this further.

Reply:
> Could there be an option that when enabled calls location_cellular_ready(NULL) immediately after the location has been handed off to the system. From the location APIs standpoint it has done its job and its up to the application to handle the ncell measurements from there.

The only drawback in this solution is that there is no knowledge if a fallback is needed. If client calls location_cellular_ready(NULL) indicating that it got ncell measurements and will handle them further, the only option in location library is to treat this as "location was retrieved successfully and no fallbacks are needed". However, resolving ncell measurements to coordinates using cloud services may fail meaning location couldn't be resolved. Location library fallback is meant for this exact purpose but then it wouldn't be made. Is this OK?

While ideologically this is far from optimal, we could implement this. As cellular is the last item in the priority list, this shouldn't be an issue in ATv2 or existing clients at the moment. However, when WiFi positioning is available and if method preference is GNSS, WiFi and cellular, fallback from WiFi to cellular would never occur if we would indicate that WiFi positioning is fine after getting list of access points. How do you think this should be handled?

Reply:
I think this is OK. It really depends on the device's scope of responsibility. I would argue that if the cloud fails to resolve the ncell measurements its the clouds problem, and not the device's. The device has carried out its job.
If the method preference is set to GNSS, WiFi and then cellular and you get wifi access points after failed GNSS, there is really no need to perform ncell measurements. But, of course if obtaining the access points list fails ncell measurements should be performed. And vice versa.

Reply:
@simensrostad I think we should think about what the end result will be, will the cloud/user get a location for the device or not.

The problem with cellular and WiFi location is, that even though the device gets the serving cell (with or without neighbors) or it finds WiFi APs, that does not mean that it can be resolved to a valid location. It's perfectly normal that there are LTE cells and WiFi APs which are not in the service provider databases. If the device gets back the result (=no location found), the Location library can perform fallback to the next method, if available, and maximize the probability for getting the device location.

Reply:
For cellular your argument about cloud failure might be true as cellular databases in cloud services are rather large although I don't believe they are 100% complete throughout the world.
But for WiFi, not being able to resolve the location is a "normal" outcome as there are so many random APs that cannot be in all cloud databases.
From this perspective, the need for fallback is very much valid.

Reply:
_But for WiFi, not being able to resolve the location is a "normal" outcome_ 😮 

Ok, I understand now why this is necessary.

Reply:
These changes for `location_cellular_ready()` works for nRF Cloud. What about AWS and Azure? Can we get the location back from these services?

If not, I guess we can get around this at the moment as cellular location is the last method and there is really no fallback from there. However, WiFi support is coming up shortly and then we would have GNSS, WiFi, cellular, and when GNSS and WiFi fails, we'd need information about WiFi failure to be able to do fallback to cellular.

Reply:
We currently don't get location back from our AWS and Azure integrations. cc @coderbyheart

Reply:
As discussed in the meeting last week, we have some alternatives here what SW does after neighbor cells are sent to cloud:

1) Cloud returns the location back, and location library is informed about it, location library does normal fallback procedures as necessary.

2) Cloud won't reply anything, application determines that location resolution is undefined and location library is informed about it. Location library will do a fallback but if it's not possible, it also determines that location resolution is undefined (LOCATION_EVT_UNDEFINED).
At the moment with this option, fallbacks won't occur as normally we have just GNSS and cellular so there is nowhere to do fallback to from cellular. But the behavior is designed to work already when WiFi is there also. Then we'd have GNSS, WiFi, cellular and we'd do fallback from WiFi to cellular always.

In this PR, ATv2 has implementation for option #2. Location library has support also for option #1.